File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/custom/type/ProductListener.java
package io.micronaut.mqtt.docs.custom.type;

// tag::imports[]
import io.micronaut.mqtt.annotation.MqttSubscriber;
import io.micronaut.mqtt.annotation.Topic;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
// end::imports[]

import io.micronaut.context.annotation.Requires;

@Requires(property = "spec.name", value = "ProductInfoSpec")
// tag::clazz[]
@MqttSubscriber
public class ProductListener {

    List<ProductInfo> messages = Collections.synchronizedList(new ArrayList<>());

    @Topic("product")
    public void receive(byte[] data,
                        ProductInfo productInfo) { // <1>
        messages.add(productInfo);
    }
}
// end::clazz[]

File: test-suite-python/src/test/python/micronaut/mqtt/docs/custom/type/ProductListener.py
# tag::imports[]
from micronaut.mqtt.annotation import MqttSubscriber, Topic

from .ProductInfo import ProductInfo
# end::imports[]

from micronaut.context.annotation import Requires


@Requires(property="spec.name", value="ProductInfoSpec")
# tag::clazz[]
@MqttSubscriber
class ProductListener:

    def __init__(self):
        self.messages: list[ProductInfo] = []

    @Topic("product")
    def receive(self, data: bytes,
                product_info: ProductInfo) -> None:  # <1>
        self.messages.append(product_info)
# end::clazz[]
